crates/rattler_repodata_gateway/src/gateway/query.rs

```diff
@@ -271,11 +271,163 @@ impl GatewayQuery {
     }
 }
 
-impl IntoFuture for GatewayQuery {
+impl IntoFuture for RepoDataQuery {
     type Output = Result<Vec<RepoData>, GatewayError>;
     type IntoFuture = futures::future::BoxFuture<'static, Self::Output>;
 
     fn into_future(self) -> Self::IntoFuture {
         self.execute().boxed()
     }
 }
+
+/// Represents a query for package names to execute with a [`Gateway`].
+///
+/// When executed the query will asynchronously load the package names from all
+/// subdirectories (combination of channels and platforms).
+///
+///
+/// Repodata is cached by the [`Gateway`] so executing the same query twice
+/// with the same channels will not result in the repodata being fetched
+/// twice.
+#[derive(Clone)]
+pub struct NamesQuery {
+    /// The gateway that manages all resources
+    gateway: Arc<GatewayInner>,
+
+    /// The channels to fetch from
+    channels: Vec<Channel>,
+
+    /// The platforms the fetch from
+    platforms: Vec<Platform>,
+
+    /// The reporter to use by the query.
+    reporter: Option<Arc<dyn Reporter>>,
+}
+
+impl NamesQuery {
+    /// Constructs a new instance. This should not be called directly, use
+    /// [`Gateway::names`] instead.
+    pub(super) fn new(
+        gateway: Arc<GatewayInner>,
+        channels: Vec<Channel>,
+        platforms: Vec<Platform>,
+    ) -> Self {
+        Self {
+            gateway,
+            channels,
+            platforms,
+
+            reporter: None,
+        }
+    }
+
+    /// Sets the reporter to use for this query.
+    ///
+    /// The reporter is notified of important evens during the execution of the
+    /// query. This allows reporting progress back to a user.
+    pub fn with_reporter(self, reporter: impl Reporter + 'static) -> Self {
+        Self {
+            reporter: Some(Arc::new(reporter)),
+            ..self
+        }
+    }
+
+    /// Execute the query and return the package names.
+    pub async fn execute(self) -> Result<Vec<String>, GatewayError> {
```

Comment:
This function is way more complex than it needs to be. There is no recursion in the names so you can do all of this concurrently. 

In pseudo code:

```
let package_names = Set::new();
for (channel, platform) in channels_and_platforms {
   subdir = get_or_create_subdir(..).await
   package_names.extend(subdir.package_names().await)
}

return package_names
```

You just need to do this concurrently for each `subdir`. Ill let you figure out how ;)

Reply:
I've removed all unnecessary parts and just used FuturesUnordered to poll results.

Let me know if it's ok